File: backend/models/widget.py
"""Embeddable sticker-builder widget models.

Four per-merchant entities that power the customer-facing widget:

- `WidgetSettings`  — one row per merchant: allowed embed origins, the shared
  webhook secret, and an optional default cutter preset.
- `PricingProfile`  — reusable pricing rules (sheet width, price per metre, gap,
  margin, per-vinyl/finish surcharges, quantity breaks). The price estimate
  reuses the gang-sheet math from the public calculator.
- `Product`         — the thing a WooCommerce/Shopify product links to. Holds the
  enabled cut styles, size limits, vinyl/finish options and bleed/safe defaults.
- `WidgetSession`   — an ephemeral design session minted from an API key + a
  product. The iframe authenticates with the session token, never the API key.

All money/rate fields are stored in the profile's `currency`. Bleed/safe are
applied automatically and never surfaced to the end customer as jargon.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from backend.models.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


class WidgetSettings(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "widget_settings"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        unique=True,
    )
    """One settings row per merchant."""

    allowed_origins: Mapped[Optional[list]] = mapped_column(JSONB, nullable=True)
    """List of origins permitted to embed the widget / call the widget API,
    e.g. ['https://shop.example.com']. Used for CORS + CSP frame-ancestors."""

    webhook_secret: Mapped[Optional[str]] = mapped_column(String(64), nullable=True)
    """Shared secret used to HMAC-verify inbound order-paid webhooks from the
    merchant's plugin. Generated on first setup; rotatable from the admin."""

    default_cutter_preset_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("cutter_presets.id", ondelete="SET NULL"),
        nullable=True,
    )

    def __repr__(self) -> str:
        return f"<WidgetSettings user={self.user_id}>"


class PricingProfile(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "pricing_profiles"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(120), nullable=False, default="Default pricing")

    currency: Mapped[str] = mapped_column(
        String(3), nullable=False, default="GBP", server_default="GBP"
    )

    sheet_width_mm: Mapped[float] = mapped_column(
        Float, nullable=False, default=300.0, server_default="300.0"
    )
    """Usable print width of the merchant's media/roll."""

    price_per_metre: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.0, server_default="0.0"
    )
    """Cost of one linear metre of media (in `currency`). Drives the sheet
    estimate: cost = price_per_metre * (sheet length used / 1000)."""

    gap_mm: Mapped[float] = mapped_column(
        Float, nullable=False, default=3.0, server_default="3.0"
    )
    """Spacing assumed between stickers when estimating how many fit."""

    margin_pct: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.0, server_default="0.0"
    )
    """Markup applied on top of media cost (e.g. 200 = +200%)."""

    handling_fee: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.0, server_default="0.0"
    )
    """Flat per-order fee added after margin (in `currency`)."""

    min_order_price: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.0, server_default="0.0"
    )
    """Price floor — the estimate is never returned below this."""

    min_length_m: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.0, server_default="0.0"
    )
    """Minimum billable length in metres. 0 = pro-rata, 1 = DTF sheets."""

    vinyl_surcharges: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    """Per-vinyl surcharge map, e.g. {'holographic': 0.15} where the value is a
    fraction added to the media cost (0.15 = +15%)."""

    finish_surcharges: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    """Per-finish surcharge map, same shape as vinyl_surcharges."""

    quantity_breaks: Mapped[Optional[list]] = mapped_column(JSONB, nullable=True)
    """Volume discounts, e.g. [{'min_qty': 50, 'discount_pct': 10},
    {'min_qty': 100, 'discount_pct': 20}] — highest matching break applies."""

    def __repr__(self) -> str:
        return f"<PricingProfile {self.name} user={self.user_id}>"


class Product(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "products"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False, default="Custom stickers")

    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )

    designer: Mapped[str] = mapped_column(
        String(16), nullable=False, default="cutout", server_default="cutout"
    )
    """Which design experience the customer gets:

    - ``cutout`` — single piece of artwork with background removal + a contour/
      face cut line (die-cut, face stickers).
    - ``canvas`` — a full multi-layer designer (text, shapes, multiple images,
      layers) on a geometric artboard (circle/oval/square/rectangle); the shape
      itself is the cut line, so no contour detection is needed.
    """

    enabled_cut_styles: Mapped[Optional[list]] = mapped_column(JSONB, nullable=True)
    """Which cut styles the customer may choose for this product, a subset of
    ['die_cut', 'face', 'square', 'circle']. Null/empty = all.

    For ``canvas`` products these double as the allowed artboard shapes
    (square/rectangle → rectangular artboard, circle → ellipse artboard)."""

    min_size_mm: Mapped[float] = mapped_column(
        Float, nullable=False, default=20.0, server_default="20.0"
    )
    max_size_mm: Mapped[float] = mapped_column(
        Float, nullable=False, default=300.0, server_default="300.0"
    )
    """Allowed sticker dimension range (applies to width and height)."""

    size_presets: Mapped[Optional[list]] = mapped_column(JSONB, nullable=True)
    """Fixed sizes the customer can pick, e.g.
    [{'width_mm': 10, 'height_mm': 10}, {'width_mm': 40, 'height_mm': 60}].
    Each preset is a width × height in mm (equal values = a square/circle
    preset). Null/empty = no presets. Used by the shaped (canvas) designer."""

    allow_custom_size: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )
    """When True the customer can also type a custom size (within min/max). When
    False and presets exist, they must pick one of the presets."""

    corner_radius: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.2, server_default="0.2"
    )
    """Default rounded-corner radius for square/rectangle artboards, as a
    fraction (0..1) of half the shorter side. 0 = sharp corners, 1 = fully
    rounded. The customer can adjust it in the designer."""

    vinyl_types: Mapped[Optional[list]] = mapped_column(JSONB, nullable=True)
    """Selectable vinyl options, e.g.
    [{'key': 'matte', 'label': 'Matte'}, {'key': 'gloss', 'label': 'Gloss'}]."""

    finishes: Mapped[Optional[list]] = mapped_column(JSONB, nullable=True)
    """Selectable finish options, same shape as vinyl_types. Null = none."""

    bleed_mm: Mapped[float] = mapped_column(
        Float, nullable=False, default=3.0, server_default="3.0"
    )
    """Applied automatically; hidden from the customer."""

    safe_mm: Mapped[float] = mapped_column(
        Float, nullable=False, default=4.0, server_default="4.0"
    )
    """Inner safe area shown only as a friendly dashed guide."""

    show_filters: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )
    """Whether to show photo filter options to customers in the widget."""

    show_ai_styles: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    """Whether to show AI style options (cartoon, pencil, etc.) in the widget."""

    show_hand_edit: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    """Whether to show the hand-edit cutline tool in the widget."""

    pricing_profile_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("pricing_profiles.id", ondelete="SET NULL"),
        nullable=True,
    )

    def __repr__(self) -> str:
        return f"<Product {self.name} user={self.user_id}>"


class WidgetSession(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "widget_sessions"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    """The merchant the session belongs to (resolved from the API key)."""

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("products.id", ondelete="CASCADE"),
        nullable=False,
    )

    token: Mapped[str] = mapped_column(String(64), nullable=False, unique=True, index=True)
    """Opaque session id embedded in the signed session token the iframe uses.
    Stored so sessions can be looked up, expired and revoked."""

    external_ref: Mapped[Optional[str]] = mapped_column(String(120), nullable=True)
    """Optional cart/line reference passed by the plugin to tie back to the store."""

    asset_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("assets.id", ondelete="SET NULL"),
        nullable=True,
    )
    """The saved sticker design once the customer finishes (single sticker)."""

    params: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    """In-progress design state / chosen options."""

    status: Mapped[str] = mapped_column(
        String(16), nullable=False, default="open", server_default="open"
    )
    """'open' | 'completed' | 'expired'."""

    expires_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )

    def __repr__(self) -> str:
        return f"<WidgetSession {self.token[:8]}… product={self.product_id}>"

File: frontend/src/api/widget.ts
import { api } from "./client";

// Cut styles available to a sticker product.
export const CUT_STYLES = [
  { key: "die_cut", label: "Die-cut (around subject)" },
  { key: "face", label: "Face cut" },
  { key: "keep_bg", label: "Keep background (cut around image)" },
  { key: "square", label: "Square / rounded" },
  { key: "circle", label: "Circle / oval" },
] as const;

export type CutStyle = (typeof CUT_STYLES)[number]["key"];

// Which cut styles each design experience allows. Cut-out products do contour
// and face cuts, plus "keep background" (a rectangle cut around the uploaded
// image, no background removal); shaped (canvas) products do geometric artboard
// shapes (rectangle/oval are reached via the in-designer unlock toggle).
export const CUT_STYLES_BY_DESIGNER: Record<"cutout" | "canvas", CutStyle[]> = {
  cutout: ["die_cut", "face", "keep_bg"],
  canvas: ["square", "circle"],
};

// ---- API keys -------------------------------------------------------------
export interface ApiKey {
  id: string;
  name: string;
  prefix: string;
  last_used_at: string | null;
  revoked_at: string | null;
  created_at: string;
}
export interface ApiKeyCreated extends ApiKey {
  plaintext: string;
}

export const listKeys = () => api<ApiKey[]>("/api/widget/keys");
export const createKey = (name: string) =>
  api<ApiKeyCreated>("/api/widget/keys", {
    method: "POST",
    body: JSON.stringify({ name }),
  });
export const revokeKey = (id: string) =>
  api<void>(`/api/widget/keys/${id}`, { method: "DELETE" });

// ---- Settings -------------------------------------------------------------
export interface WidgetSettings {
  allowed_origins: string[];
  has_webhook_secret: boolean;
}
export const getWidgetSettings = () => api<WidgetSettings>("/api/widget/settings");
export const updateWidgetSettings = (allowed_origins: string[]) =>
  api<WidgetSettings>("/api/widget/settings", {
    method: "PATCH",
    body: JSON.stringify({ allowed_origins }),
  });
export const rotateWebhookSecret = () =>
  api<{ webhook_secret: string }>("/api/widget/settings/webhook-secret", {
    method: "POST",
  });

// ---- Pricing profiles -----------------------------------------------------
export interface QuantityBreak {
  min_qty: number;
  discount_pct: number;
}
export interface PricingProfile {
  id: string;
  name: string;
  currency: string;
  sheet_width_mm: number;
  price_per_metre: number;
  gap_mm: number;
  margin_pct: number;
  handling_fee: number;
  min_order_price: number;
  min_length_m: number;
  vinyl_surcharges: Record<string, number> | null;
  finish_surcharges: Record<string, number> | null;
  quantity_breaks: QuantityBreak[] | null;
  quantity_presets: number[] | null;
  allow_custom_quantity: boolean;
  created_at: string;
}
export type PricingProfileInput = Omit<PricingProfile, "id" | "created_at">;

export const listPricingProfiles = () =>
  api<PricingProfile[]>("/api/widget/pricing-profiles");
export const createPricingProfile = (body: PricingProfileInput) =>
  api<PricingProfile>("/api/widget/pricing-profiles", {
    method: "POST",
    body: JSON.stringify(body),
  });
export const updatePricingProfile = (id: string, body: PricingProfileInput) =>
  api<PricingProfile>(`/api/widget/pricing-profiles/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
export const deletePricingProfile = (id: string) =>
  api<void>(`/api/widget/pricing-profiles/${id}`, { method: "DELETE" });

// ---- Products -------------------------------------------------------------
export interface VinylOption {
  key: string;
  label: string;
}
export type DesignerMode = "cutout" | "canvas";

export interface SizePreset {
  width_mm: number;
  height_mm: number;
}

export interface Product {
  id: string;
  name: string;
  is_active: boolean;
  designer: DesignerMode;
  enabled_cut_styles: CutStyle[];
  min_size_mm: number;
  max_size_mm: number;
  size_presets: SizePreset[];
  allow_custom_size: boolean;
  corner_radius: number;
  vinyl_types: VinylOption[];
  finishes: VinylOption[];
  bleed_mm: number;
  safe_mm: number;
  show_filters: boolean;
  show_ai_styles: boolean;
  show_hand_edit: boolean;
  require_proof: boolean;
  proof_fee: number;
  pricing_profile_id: string | null;
  created_at: string;
}
export type ProductInput = Omit<Product, "id" | "created_at">;

export const listProducts = () => api<Product[]>("/api/widget/products");
export const getProduct = (id: string) => api<Product>(`/api/widget/products/${id}`);
export const createProduct = (body: ProductInput) =>
  api<Product>("/api/widget/products", {
    method: "POST",
    body: JSON.stringify(body),
  });
export const updateProduct = (id: string, body: ProductInput) =>
  api<Product>(`/api/widget/products/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
export const deleteProduct = (id: string) =>
  api<void>(`/api/widget/products/${id}`, { method: "DELETE" });

// ---- Print orders ---------------------------------------------------------
export interface PrintOrder {
  id: string;
  platform: string;
  external_order_id: string;
  customer_ref: string | null;
  line_items: Array<Record<string, unknown>>;
  amount_total: number;
  currency: string;
  status: "draft" | "paid" | "ready_to_print" | "printed";
  proof_status: string | null;
  proof_notes: string | null;
  proof_history: Array<{ action: string; timestamp: string; by: string; message: string }> | null;
  customer_email: string | null;
  proof_token: string | null;
  output_r2_key: string | null;
  created_at: string;
}
export const listOrders = (status?: string) =>
  api<PrintOrder[]>(
    `/api/widget/orders${status ? `?status_filter=${encodeURIComponent(status)}` : ""}`
  );
export const updateOrderStatus = (id: string, status: PrintOrder["status"]) =>
  api<PrintOrder>(`/api/widget/orders/${id}`, {
    method: "PATCH",
    body: JSON.stringify({ status }),
  });
export const deleteOrder = (id: string) =>
  api<void>(`/api/widget/orders/${id}`, { method: "DELETE" });

export const sendProof = (id: string) =>
  api<PrintOrder>(`/api/widget/orders/${id}/send-proof`, { method: "POST" });

// ---- Live preview ---------------------------------------------------------
export const createPreviewSession = (product_id: string) =>
  api<{ session_token: string; expires_in: number }>("/api/widget/preview-session", {
    method: "POST",
    body: JSON.stringify({ product_id }),
  });
